### PR TITLE
vod: Create catalyst pipeline coordinator

### PR DESCRIPTION
**Pull request by victorges**
This is to create the most basic abstraction that I plan on using for the MediaConvert pipeline.

The idea is that this new component will centralize all the business logic related to running the VOD
jobs, be it one in foreground another in background, using as a fallback, etc. The API handlers should
only handle the protocols and schemas of the APIs, then just pass the requests to the `pipeline.Coordinator`
which will coordinate the configured pipelines for the jobs.

The process for implementing this was basically checking all the usages of the `StreamCache` and move
those code paths into this new `pipeline.Coordinator` system (which is the only one with access to the 
whole cache now). Also abstracted the status callback client so that a fake one can be sent to the jobs
running in background.

After that, created a `pipeline.Handler` which has implementations an existing implementation for Mist, which
was fully copied and then adapted from the existing implementation, and one stub implementation for 
AWS MediaConvert (today it only errors on every attempt). 

The pipeline coordinator itself remained responsible only for managing the background go-routines and 
passing of messages between the workflows, the callback clients etc. Its main responsibilities are:
 - handle running the jobs in background (e.g. here we'll be able to know "how many jobs are running now"?)
 - handle correct completion of the jobs, making sure to send an error callback on any fatal flow
 - coordinate when to run jobs in Mist and when to run them in MediaConvert (as well as fallback)

The only missing piece to abstract to get to that will be the callbacks. Those should probably be tied to the workflow engine state as well, and we should only actually send the callbacks for the foreground operation, not the background one (on the duplication phase).

This is now with full documentation and tests on the main components created. Ready for review!

### CODE REVIEW

**Review by thomshutt — APPROVED** [review]
Discussed some of the indirection offline with Victor, but overall this looks great!

File: pipeline/coordinator.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+	Jobs *cache.Cache[*JobInfo]
+}
+
+// This has the main logic regarding the pipeline strategy. It starts jobs and
+// handles processing the response and triggering a fallback if appropriate.
+func (c *coord) StartUploadJob(p UploadJobPayload) {
+	switch c.strategy {
+	case StrategyCatalystDominance:
+		c.startOneUploadJob(p, c.pipeMist, true)
```
[the remark this answers is not in the document]
Comment (victorges) [reply pipeline/coordinator.go:139]: Added another log just to make sure we log everything we send on this channel

File: pipeline/coordinator.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	StreamName   string
+	ReportStatus TranscodeStatusReporter
+
+	handler Handler
+}
+
+// Coordinator is the main interface to handle the pipelines. It should be
+// called directly from the API handlers and never blocks on execution but
+// rather schedules some routines in background.
+type Coordinator interface {
```
[the remark this answers is not in the document]
Comment (victorges) [reply pipeline/coordinator.go:80]: IMO the advantage of the interface is concise documentation of the available methods, which is generally better than trying to parse through a long file looking for public functions. Also helps to document here only the parts that are publicly relevant, while including implementation details in the docs of the actual implementation.

This is a very soft preference on the style tho, and I think it's better to be consistent in the project anyway, so I'll remove it and move the docs.

File: handlers/misttriggers/triggers.go [hunk 3]
```diff
@@ -69,7 +69,7 @@ func streamNameToPipeline(name string) PipelineId {
 	return Unrelated
 }
 
-type PipelineId = int
```
[the remark this answers is not in the document]
Comment (victorges) [reply handlers/misttriggers/triggers.go:72]: Yeah it's confusing. It's a typedef vs an actual runtime type

**Review by thomshutt — APPROVED** [review]

File: pipeline/mist.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+			log.LogError(requestID, "Failed to delete stream in triggerRecordingEndSegmenting", err)
+		}
+	}()
+
+	// Let Studio know that we've almost finished the Segmenting phase
+	job.ReportStatus(clients.NewTranscodeStatusProgress(callbackUrl, requestID, clients.TranscodeStatusPreparing, 0.9))
+
+	// HACK: Wait a little bit to give the segmenting time to finish uploading.
+	// Proper fix comes with a new Mist trigger to notify us that uploads are also complete
+	time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (red-0ne) [reply pipeline/mist.go:90]: Yes should be safe to be removed. But not sure if the timing configured will be enough.

**Pull request merged**